assignment-client/src/assets/AssetServer.cpp

```diff
@@ -0,0 +1,279 @@
+                continue;
+            }
+
+            // Read file
+            QFile file { fileInfo.absoluteFilePath() };
+            file.open(QFile::ReadOnly);
+            QByteArray data = file.readAll();
+
+            auto hash = hashData(data);
+
+            qDebug() << "\tMoving " << filename << " to " << hash;
+
+            file.rename(_resourcesDirectory.absoluteFilePath(hash));
+        }
+    }
+
+    while (!_isFinished) {
+        // since we're a while loop we need to help Qt's event processing
+        QCoreApplication::processEvents();
+        // QCoreApplication::processEvents(QEventLoop::WaitForMoreEvents);
+    }
+}
+
+void AssetServer::handleAssetGetInfo(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    if (packet->getPayloadSize() < HASH_HEX_LENGTH) {
+        qDebug() << "ERROR bad file request";
+        return;
+    }
+
+    QByteArray assetHash;
+    MessageID messageID;
+    packet->readPrimitive(&messageID);
+    assetHash = packet->read(HASH_HEX_LENGTH);
+
+    auto replyPacket = NLPacket::create(PacketType::AssetGetInfoReply);
+
+    replyPacket->writePrimitive(messageID);
+    replyPacket->write(assetHash, HASH_HEX_LENGTH);
+
+    QFileInfo fileInfo { _resourcesDirectory.filePath(QString(assetHash)) };
+    qDebug() << "Opening file: " << QString(QFileInfo(assetHash).fileName());
+
+    if (fileInfo.exists() && fileInfo.isReadable()) {
+        replyPacket->writePrimitive(AssetServerError::NO_ERROR);
+        replyPacket->writePrimitive(fileInfo.size());
+    } else {
+        qDebug() << "Asset not found: " << assetHash;
+        replyPacket->writePrimitive(AssetServerError::ASSET_NOT_FOUND);
+    }
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->sendPacket(std::move(replyPacket), *senderNode);
+}
+
+void AssetServer::handleAssetGet(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    if (packet->getPayloadSize() < HASH_HEX_LENGTH) {
+        qDebug() << "ERROR bad file request";
+        return;
+    }
+
+    MessageID messageID;
+    packet->readPrimitive(&messageID);
+
+    QByteArray assetHash;
+    assetHash = packet->read(HASH_HEX_LENGTH);
+
+    DataOffset start;
+    packet->readPrimitive(&start);
+
+    DataOffset end;
+    packet->readPrimitive(&end);
+
+    qDebug() << "Received a request for the file: " << assetHash << " from " << start << " to " << end;
+
+    // Queue task
+    QString filePath = _resourcesDirectory.filePath(QString(assetHash));
+    auto task = new SendAssetTask(messageID, assetHash, filePath, start, end, senderNode);
+    _taskPool.start(task);
+}
+
+void AssetServer::handleAssetUpload(QSharedPointer<NLPacketList> packetList, SharedNodePointer senderNode) {
+    auto data = packetList->getMessage();
+    QBuffer buffer { &data };
+    buffer.open(QIODevice::ReadOnly);
+
+    MessageID messageID;
+    buffer.read(reinterpret_cast<char*>(&messageID), sizeof(messageID));
+    // buffer.readPrimitive(&messageID);
```

Comment:
Yeah - I thought about adding helper methods for this `readPrimitive<T>(QIODevice, T)`, but I figured we would be changing this altogether in the nearish future.